cmd/klotho/main.go

```diff
@@ -219,7 +220,16 @@ func run(cmd *cobra.Command, args []string) (err error) {
 
 	if appCfg.AppName == "" {
 		return errors.New("'app' required")
+	} else if len(appCfg.AppName) > 20 {
+		return fmt.Errorf("'app' must be less than 20 characters in length. 'app' was %s", cfg.appName)
 	}
+	match, err := regexp.MatchString(`^[A-Za-z0-9-_]+$`, cfg.appName)
```

Comment:
nit: `^[\w-]+$` would be a bit more concise

Reply:
i can change it. Also do we think 20 characters is too small for these?

Reply:
That's kind of what I was getting at with having to find a pretty restrictive intersection of all the constraints. I think (a) 20 feels on the small side from a usability perspective, but also (b) 20 feels about right from a useful-as-a-length-constraint perspective.

🤷 

I guess we can put it in and see if people complain. Maybe also add a log line when it happens, so we can get metrics on how often that is?

Reply:
the only issue with a log line is then its visible to the user in both an error sense and logging sense right? The only method we have of streaming back to data dog is through the zap logger.

Reply:
Oh hm, actually, the fact that there's an error will already get reported back.  Let me check if that contains the error message; if so, we shouldn't do that (since it's potentially sensitive), but it'd be good to include it as a sanitized field (maybe a new kind called "Silent" that gets ignored in the stderr output?).

Reply:
Okay yeah, we don't send the log line (whew!). So, I would:

1. create a new struct in `logging/fields.go` called `silentAnalytics { key string, data any }` or similar , along with a `SilentAnalytics` function (you can look at `annotationField` for an example)
2. add an instance of that to this log line
3. I think you don't need to do anything else; it'll get reported in analytics, but won't get logged extra to console

Reply:
I would like to have `.`, `:`, and maybe `/` added to the allowable characters as well as those are common separators.

Reply:
We can also move the length checking to the regexp `^[\w-.:/]{3,20}$`. And if you compile it first, you can use the compiled `.String()` to print its regex in the error message.

Reply:
> We can also move the length checking to the regexp `^[\w-.:/]{3,20}$`

I like it as it is now, because it gives a more precise indication of the problem. That is, "can only contain this set of chars" and "has to be 3-20 chars long" are separate enough that I'd prefer (as a user) to be told which of them I'm violating, even though it's true that a single regex can check both.

Reply:
added the characters requested and added an analytics Error. Verified in datadog we see

```
[User: jhsinger2+1@gmail.com] Klotho parameter check failed. 'app' can only contain alphanumeric, -, _, ., :, and /.
```